server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java

```diff
@@ -147,114 +157,170 @@ public MetadataIndexStateService(
      * to be terminated and finally closes the indices by moving their state to CLOSE.
      */
     public void closeIndices(final CloseIndexClusterStateUpdateRequest request, final ActionListener<CloseIndexResponse> listener) {
-        final Index[] concreteIndices = request.indices();
-        if (concreteIndices == null || concreteIndices.length == 0) {
+        if (request.indices() == null || request.indices().length == 0) {
             throw new IllegalArgumentException("Index name is required");
         }
 
         clusterService.submitStateUpdateTask(
-            "add-block-index-to-close " + Arrays.toString(concreteIndices),
-            new ClusterStateUpdateTask(Priority.URGENT, request.masterNodeTimeout()) {
+            "add-block-index-to-close " + Arrays.toString(request.indices()),
+            new AddBlocksToCloseTask(request, listener),
+            ClusterStateTaskConfig.build(Priority.URGENT, request.masterNodeTimeout()),
+            this.addBlocksToCloseExecutor
+        );
+    }
 
-                private final Map<Index, ClusterBlock> blockedIndices = new HashMap<>();
+    private class AddBlocksToCloseExecutor implements ClusterStateTaskExecutor<AddBlocksToCloseTask> {
 
-                @Override
-                public ClusterState execute(final ClusterState currentState) {
-                    return addIndexClosedBlocks(concreteIndices, blockedIndices, currentState);
+        @Override
+        public ClusterTasksResult<AddBlocksToCloseTask> execute(ClusterState currentState, List<AddBlocksToCloseTask> tasks)
+            throws Exception {
+            ClusterTasksResult.Builder<AddBlocksToCloseTask> builder = ClusterTasksResult.builder();
+            ClusterState state = currentState;
+
+            for (AddBlocksToCloseTask task : tasks) {
+                try {
+                    final Map<Index, ClusterBlock> blockedIndices = new HashMap<>(task.request.indices().length);
+                    state = addIndexClosedBlocks(task.request.indices(), blockedIndices, state);
+                    builder.success(task, new ActionListener<>() {
+                        @Override
+                        public void onResponse(ClusterState clusterState) {
+                            if (blockedIndices.isEmpty()) {
+                                task.listener.onResponse(new CloseIndexResponse(true, false, Collections.emptyList()));
+                            } else {
+                                threadPool.executor(ThreadPool.Names.MANAGEMENT)
```

Comment:
I don't see a great reason to fork here, we really just trigger a bunch of transport actions that themselves fork to `MANAGEMENT` again. But then again this is sending `O(#shards)` transport actions in parallel, which could be lots, and we know that can cause problems. No change needed here in this PR but we might want to do something about that in future perhaps.